### PR TITLE
refactor(source): extract parameters required by each parser

### PR DESCRIPTION
**Pull request by wugouzi**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
..so that we can avoid using hashmap of properties to create parser.

We separate encoding config and protocol config in this pr from properties that are required to build parser.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

## Documentation

- [ ] My PR contains user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/connector/src/parser/avro/parser.rs [hunk 1]
```diff
@@ -54,20 +54,29 @@ pub struct AvroParserConfig {
...
 
 impl AvroParserConfig {
-    pub async fn new(
-        props: &HashMap<String, String>,
-        schema_location: &str,
-        use_schema_registry: bool,
-        enable_upsert: bool,
-        upsert_primary_key_column_name: Option<String>,
-    ) -> Result<Self> {
+    pub async fn new(parser_config_list: &ParserConfigList, enable_upsert: bool) -> Result<Self> {
```
Comment (tabVersion) [inline src/connector/src/parser/avro/parser.rs:57]: Consider derive enum_as_inner for ParserConfigList

File: src/connector/src/parser/mod.rs [hunk 2]
```diff
@@ -493,6 +495,142 @@ pub enum SpecificParserConfig {
     Bytes,
 }
 
+#[derive(Debug, Clone, Default)]
+pub struct ParserClientConfigList {
+    username: Option<String>,
+    password: Option<String>,
```
Comment (tabVersion) [inline src/connector/src/parser/mod.rs:501]: Is it just for schema registry?

File: src/connector/src/parser/mod.rs [hunk 3]
```diff
@@ -493,6 +495,142 @@ pub enum SpecificParserConfig {
...
+}
+
+#[derive(Default)]
+pub struct AvroParserConfigList {
+    pub use_schema_registry: bool,
+    pub row_schema_location: String,
+    pub aws_auth_props: Option<AwsAuthProps>,
+    pub topic: String,
+    pub client_config: ParserClientConfigList,
+    pub primary_key: String,
```
Comment (tabVersion) [inline src/connector/src/parser/mod.rs:529]: Consider using pk_indices 'vec<int>' instead

File: src/connector/src/parser/mod.rs [hunk 4]
```diff
@@ -493,6 +495,142 @@ pub enum SpecificParserConfig {
...
+    pub message_name: String,
+    pub use_schema_registry: bool,
+    pub row_schema_location: String,
+    pub aws_auth_props: Option<AwsAuthProps>,
+    pub topic: String,
+    pub client_config: ParserClientConfigList,
+}
+
+#[derive(Default)]
+pub struct DebeziumAvroParserConfigList {
```
Comment (tabVersion) [inline src/connector/src/parser/mod.rs:543]: We should not concat Avro and debezium into one config. 
It is expected to do the same with separate configs.

File: src/connector/src/parser/mod.rs (hunk 2 above)
Comment (wugouzi) [reply src/connector/src/parser/mod.rs:501]: Yes

File: src/connector/src/parser/mod.rs (hunk 4 above)
Comment (wugouzi) [reply src/connector/src/parser/mod.rs:543]: This is for Debezium avro only

File: src/connector/src/parser/csv_parser.rs [hunk 5]
```diff
@@ -36,6 +36,23 @@ pub struct CsvParserConfig {
...
+impl CsvParserConfig {
+    pub fn new(parser_properties: ParserProperties) -> Result<Self> {
+        let csv_config = if let EncodingProperties::Csv(config) = parser_properties.encoding_config
+        {
+            config
+        } else {
+            return Err(RwError::from(ProtocolError(
+                "wrong parser config list for Csv".to_string(),
+            )));
+        };
```
Comment (tabVersion) [inline src/connector/src/parser/csv_parser.rs:48]: may `try_match_expand!` here?

File: src/connector/src/parser/debezium/avro_parser.rs [hunk 6]
```diff
@@ -57,14 +58,24 @@ pub struct DebeziumAvroParserConfig {
...
-    pub async fn new(props: &HashMap<String, String>, schema_location: &str) -> Result<Self> {
+    pub async fn new(parser_properties: ParserProperties) -> Result<Self> {
+        let avro_config =
+            if let EncodingProperties::Avro(config) = parser_properties.encoding_config {
+                config
+            } else {
+                return Err(RwError::from(ProtocolError(
+                    "wrong parser config list for Avro".to_string(),
+                )));
+            };
```
Comment (tabVersion) [inline src/connector/src/parser/debezium/avro_parser.rs:69]: ditto

File: src/connector/src/parser/mod.rs (hunk 2 above)
Comment (tabVersion) [reply src/connector/src/parser/mod.rs:501]: I think better name it `SchemaRegistryAuth`

File: src/connector/src/parser/mod.rs [hunk 7]
```diff
@@ -493,6 +495,147 @@ pub enum SpecificParserConfig {
...
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct AvroProperties {
+    pub use_schema_registry: bool,
+    pub row_schema_location: String,
+    pub upsert_primary_key: String,
+    pub client_config: ParserClientConfigList,
```
Comment (tabVersion) [inline src/connector/src/parser/mod.rs:521]: which client?

File: src/connector/src/parser/mod.rs [hunk 8]
```diff
@@ -493,6 +495,147 @@ pub enum SpecificParserConfig {
...
+    pub has_header: bool,
+}
+
+#[derive(Default)]
+pub enum EncodingProperties {
+    Avro(AvroProperties),
+    Protobuf(ProtobufProperties),
+    Csv(CsvProperties),
+    #[default]
+    None,
```
Comment (tabVersion) [inline src/connector/src/parser/mod.rs:549]: lack `bytes` encode?

File: src/connector/src/parser/mod.rs [hunk 9]
```diff
@@ -493,6 +495,147 @@ pub enum SpecificParserConfig {
...
+///    client info, topic, aws config, optional primary key
+/// 3. PROTOBUF: message name, row schema location, use schema registry
+///    client info, topic, aws config
+/// 4. DEBEZIUM AVRO: row schema location, topic, client info
+/// 5. Other: none
+
+impl ParserProperties {
+    pub fn new(
+        format: SourceFormat,
+        props: &HashMap<String, String>,
```
Comment (tabVersion) [inline src/connector/src/parser/mod.rs:578]: can we access fields in `props` by `remove`? and report error if there is anything left when returning

File: src/connector/src/parser/mod.rs [hunk 10]
```diff
@@ -526,44 +669,20 @@ impl SpecificParserConfig {
...
+            }
+            SourceFormat::Avro => {
+                SpecificParserConfig::Avro(AvroParserConfig::new(parser_properties).await?)
+            }
+            SourceFormat::UpsertAvro => {
+                SpecificParserConfig::UpsertAvro(AvroParserConfig::new(parser_properties).await?)
+            }
+            SourceFormat::Protobuf => {
+                SpecificParserConfig::Protobuf(ProtobufParserConfig::new(parser_properties).await?)
+            }
```
Comment (tabVersion) [inline src/connector/src/parser/mod.rs:685]: replace with macro here?

File: src/connector/src/parser/protobuf/parser.rs [hunk 11]
```diff
@@ -49,20 +50,24 @@ pub struct ProtobufParserConfig {
...
-    ) -> Result<Self> {
+    pub async fn new(parser_properties: ParserProperties) -> Result<Self> {
+        let protobuf_config =
+            if let EncodingProperties::Protobuf(config) = parser_properties.encoding_config {
+                config
+            } else {
+                return Err(RwError::from(ProtocolError(
+                    "wrong parser config list for Protobuf".to_string(),
+                )));
+            };
```
Comment (tabVersion) [inline src/connector/src/parser/protobuf/parser.rs:61]: ditto

**Review by tabVersion — APPROVED** [review]
rest lgtm, please also resolve conflicts

File: src/connector/src/parser/mod.rs (hunk 7 above)
Comment (wugouzi) [reply src/connector/src/parser/mod.rs:521]: it is used to create `Client` for schema resolver/schema registry

File: src/connector/src/parser/mod.rs (hunk 9 above)
Comment (wugouzi) [reply src/connector/src/parser/mod.rs:578]: I think we can't. Currently we extract topic, `SchemaRegistryAuth`, `AwsAuthProps` from props and topic will be used later iiuc. For the auth part, it's hard to tell if any field is missing.

File: src/connector/src/parser/mod.rs (hunk 10 above)
Comment (wugouzi) [reply src/connector/src/parser/mod.rs:685]: I think we can wait until api is stable. I am not sure yet

**Review by st1page — APPROVED** [review]

**Pull request merged**